File: src/main/java/com/wikia/tibia/objects/Item.java
package com.wikia.tibia.objects;

import com.fasterxml.jackson.annotation.JsonGetter;
import com.wikia.tibia.enums.Article;
import com.wikia.tibia.enums.DamageElement;
import com.wikia.tibia.enums.Hands;
import com.wikia.tibia.enums.ItemClass;
import com.wikia.tibia.enums.Status;
import com.wikia.tibia.enums.WeaponType;
import com.wikia.tibia.enums.YesNo;
import com.wikia.tibia.utils.ObjectUtils;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.math.BigDecimal;
import java.util.List;

@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class Item extends WikiObject {

    private static final Logger LOG = LoggerFactory.getLogger(Item.class);

    private List<Integer> itemid;
    private YesNo marketable;
    private YesNo usable;
    private String sprites;
    private String flavortext;
    private Status ingamestatus;
    private String words;
    private ItemClass itemclass;
    private String primarytype;
    private String secondarytype;
    private Integer lightcolor;
    private Integer lightradius;
    private Integer levelrequired;
    private String vocrequired;
    private Integer mlrequired;
    private Hands hands;
    private WeaponType type;
    private String attack;
    private Integer fireAttack;
    private Integer earthAttack;
    private Integer iceAttack;
    private Integer energyAttack;
    private Integer defense;
    private String defensemod;
    private Integer imbueslots;
    private String imbuements;
    private YesNo enchantable;
    private YesNo enchanted;
    private String range;
    private String attackModification;
    private String hitpointModification;
    private Integer armor;
    private String resist;
    private Integer charges;
    private Percentage criticalHitChance;
    private Percentage criticalHitExtraDamage;
    private Percentage manaleechChance;
    private Percentage manaleechAmount;
    private Percentage hitpointLeechChance;
    private Percentage hitpointLeechAmount;
    private String attrib;
    private BigDecimal weight;
    private YesNo stackable;
    private YesNo pickupable;
    private YesNo immobile;
    private YesNo walkable;
    private YesNo unshootable;
    private YesNo blockspath;
    private YesNo rotatable;
    private Integer mapcolor;
    private YesNo consumable;
    private Integer regenseconds;
    private List<String> sounds;
    private YesNo writable;
    private YesNo rewritable;
    private Integer writechars;
    private YesNo hangable;
    private YesNo holdsliquid;
    private Integer mana;
    private DamageElement damagetype;
    private String damage;
    private Integer volume;
    private String duration;
    private YesNo destructible;
    private List<String> droppedby;
    private String value;
    private String npcvalue;
    private String npcprice;
    private String npcvaluerook;
    private String npcpricerook;
    private String buyfrom;
    private String sellto;

    @Builder
    private Item(String name, Article article, String actualname, String plural, String implemented, String notes,
                 String history, Status status, List<Integer> itemid, YesNo marketable, YesNo usable, String sprites,
                 String flavortext, Status ingamestatus, String words, ItemClass itemclass, String primarytype,
                 String secondarytype, Integer lightcolor, Integer lightradius, Integer levelrequired,
                 String vocrequired, Integer mlrequired, Hands hands, WeaponType type, String attack,
                 Integer fireAttack, Integer earthAttack, Integer iceAttack, Integer energyAttack, Integer defense,
                 String defensemod, Integer imbueslots, String imbuements,
                 YesNo enchantable, YesNo enchanted, String range, String attackModification, String hitpointModification,
                 Integer armor, String resist, Integer charges, Percentage criticalHitChance,
                 Percentage criticalHitExtraDamage, Percentage manaleechChance, Percentage manaleechAmount,
                 Percentage hitpointLeechChance, Percentage hitpointLeechAmount, String attrib, BigDecimal weight,
                 YesNo stackable, YesNo pickupable, YesNo immobile, YesNo walkable, YesNo unshootable, YesNo blockspath,
                 YesNo rotatable, Integer mapcolor, YesNo consumable, Integer regenseconds, List<String> sounds,
                 YesNo writable, YesNo rewritable, Integer writechars, YesNo hangable, YesNo holdsliquid, Integer mana,
                 DamageElement damagetype, String damage, Integer volume, String duration, YesNo destructible,
                 List<String> droppedby, String value, String npcvalue, String npcprice, String npcvaluerook,
                 String npcpricerook, String buyfrom, String sellto) {
        super(name, article, actualname, plural, implemented, notes, history, status);
        this.itemid = itemid;
        this.marketable = marketable;
        this.usable = usable;
        this.sprites = sprites;
        this.flavortext = flavortext;
        this.ingamestatus = ingamestatus;
        this.words = words;
        this.itemclass = itemclass;
        this.primarytype = primarytype;
        this.secondarytype = secondarytype;
        this.lightcolor = lightcolor;
        this.lightradius = lightradius;
        this.levelrequired = levelrequired;
        this.vocrequired = vocrequired;
        this.mlrequired = mlrequired;
        this.hands = hands;
        this.type = type;
        this.attack = attack;
        this.fireAttack = fireAttack;
        this.earthAttack = earthAttack;
        this.iceAttack = iceAttack;
        this.energyAttack = energyAttack;
        this.defense = defense;
        this.defensemod = defensemod;
        this.imbueslots = imbueslots;
        this.imbuements = imbuements;
        this.enchantable = enchantable;
        this.enchanted = enchanted;
        this.range = range;
        this.attackModification = attackModification;
        this.hitpointModification = hitpointModification;
        this.armor = armor;
        this.resist = resist;
        this.charges = charges;
        this.criticalHitChance = criticalHitChance;
        this.criticalHitExtraDamage = criticalHitExtraDamage;
        this.manaleechChance = manaleechChance;
        this.manaleechAmount = manaleechAmount;
        this.hitpointLeechChance = hitpointLeechChance;
        this.hitpointLeechAmount = hitpointLeechAmount;
        this.attrib = attrib;
        this.weight = weight;
        this.stackable = stackable;
        this.pickupable = pickupable;
        this.immobile = immobile;
        this.walkable = walkable;
        this.unshootable = unshootable;
        this.blockspath = blockspath;
        this.rotatable = rotatable;
        this.mapcolor = mapcolor;
        this.consumable = consumable;
        this.regenseconds = regenseconds;
        this.sounds = sounds;
        this.writable = writable;
        this.rewritable = rewritable;
        this.writechars = writechars;
        this.hangable = hangable;
        this.holdsliquid = holdsliquid;
        this.mana = mana;
        this.damagetype = damagetype;
        this.damage = damage;
        this.volume = volume;
        this.duration = duration;
        this.destructible = destructible;
        this.droppedby = droppedby;
        this.value = value;
        this.npcvalue = npcvalue;
        this.npcprice = npcprice;
        this.npcvaluerook = npcvaluerook;
        this.npcpricerook = npcpricerook;
        this.buyfrom = buyfrom;
        this.sellto = sellto;
    }

    @JsonGetter("atk_mod")
    public String getAttackModification() {
        return attackModification;
    }

    @JsonGetter("hit_mod")
    public String getHitpointModification() {
        return hitpointModification;
    }

    @JsonGetter("crithit_ch")
    public Percentage getCriticalHitChance() {
        return criticalHitChance;
    }

    @JsonGetter("critextra_dmg")
    public Percentage getCriticalHitExtraDamage() {
        return criticalHitExtraDamage;
    }

    @JsonGetter("manaleech_ch")
    public Percentage getManaleechChance() {
        return manaleechChance;
    }

    @JsonGetter("manaleech_am")
    public Percentage getManaleechAmount() {
        return manaleechAmount;
    }

    @JsonGetter("hpleech_ch")
    public Percentage getHitpointLeechChance() {
        return hitpointLeechChance;
    }

    @JsonGetter("hpleech_am")
    public Percentage getHitpointLeechAmount() {
        return hitpointLeechAmount;
    }

    @JsonGetter("fire_attack")
    public Integer getFireAttack() {
        return fireAttack;
    }

    @JsonGetter("earth_attack")
    public Integer getEarthAttack() {
        return earthAttack;
    }

    @JsonGetter("ice_attack")
    public Integer getIceAttack() {
        return iceAttack;
    }

    @JsonGetter("energy_attack")
    public Integer getEnergyAttack() {
        return energyAttack;
    }

    @Override
    public void setDefaultValues() {

        if (ObjectUtils.isEmpty(getImplemented())) {
            setImplemented("?");
        }

        if (ObjectUtils.isEmpty(itemclass)) {
            LOG.warn("Creature '{}' has no itemclass set", getName());
        }

        if (ObjectUtils.isEmpty(primarytype)) {
            primarytype = "?";
            LOG.warn("Creature '{}' has no primarytype set", getName());
        }

        if (ObjectUtils.isEmpty(value)) {
            value = "?";
        }

        if (ObjectUtils.isEmpty(npcvalue)) {
            npcvalue = "?";
        }

        if (ObjectUtils.isEmpty(npcprice)) {
            npcprice = "?";
        }

        if (ObjectUtils.isEmpty(buyfrom)) {
            buyfrom = "?";
        }

        if (ObjectUtils.isEmpty(sellto)) {
            sellto = "?";
        }
    }
}

File: src/main/kotlin/com/wikia/tibia/objects/Item.kt
package com.wikia.tibia.objects

import com.fasterxml.jackson.annotation.JsonGetter
import com.wikia.tibia.enums.*
import com.wikia.tibia.utils.ObjectUtils.isEmpty
import org.slf4j.LoggerFactory
import java.math.BigDecimal

data class Item(
        val name: String,
        val article: Article? = null,
        val actualname: String,
        val plural: String? = null,
        val implemented: String? = null,
        val notes: String? = null,
        val notes2: String? = null,
        val history: String? = null,
        val status: Status? = null,
        val itemid: List<Int>? = emptyList(),
        val marketable: YesNo? = null,
        val usable: YesNo? = null,
        val sprites: String? = null,
        val flavortext: String? = null,
        val ingamestatus: Status? = null,
        val words: String? = null,
        val itemclass: ItemClass,
        val primarytype: String? = null,
        val secondarytype: String? = null,
        val lightcolor: Int? = null,
        val lightradius: Int? = null,
        val levelrequired: Int? = null,
        val vocrequired: String? = null,
        val mlrequired: Int? = null,
        val hands: Hands? = null,
        val type: WeaponType? = null,
        val attack: String? = null,
        @get:JsonGetter("fire_attack") val fireAttack: Int? = null,
        @get:JsonGetter("earth_attack") val earthAttack: Int? = null,
        @get:JsonGetter("ice_attack") val iceAttack: Int? = null,
        @get:JsonGetter("energy_attack") val energyAttack: Int? = null,
        @get:JsonGetter("death_attack") val deathAttack: Int? = null,
        val defense: Int? = null,
        val defensemod: String? = null,
        val imbueslots: Int? = null,
        val imbuements: String? = null,
        val enchantable: YesNo? = null,
        val enchanted: YesNo? = null,
        val range: String? = null,
        @get:JsonGetter("atk_mod") val attackModification: String? = null,
        @get:JsonGetter("hit_mod") val hitpointModification: String? = null,
        val armor: Int? = null,
        val resist: String? = null,
        val charges: Int? = null,
        @get:JsonGetter("crithit_ch") val criticalHitChance: Percentage? = null,
        @get:JsonGetter("critextra_dmg") val criticalHitExtraDamage: Percentage? = null,
        @get:JsonGetter("manaleech_ch") val manaleechChance: Percentage? = null,
        @get:JsonGetter("manaleech_am") val manaleechAmount: Percentage? = null,
        @get:JsonGetter("hpleech_ch") val hitpointLeechChance: Percentage? = null,
        @get:JsonGetter("hpleech_am") val hitpointLeechAmount: Percentage? = null,
        val attrib: String? = null,
        val weight: BigDecimal? = null,
        val stackable: YesNo? = null,
        val pickupable: YesNo? = null,
        val immobile: YesNo? = null,
        val walkable: YesNo? = null,
        val unshootable: YesNo? = null,
        val blockspath: YesNo? = null,
        val rotatable: YesNo? = null,
        val mapcolor: Int? = null,
        val consumable: YesNo? = null,
        val regenseconds: Int? = null,
        val sounds: List<String>? = emptyList(),
        val writable: YesNo? = null,
        val rewritable: YesNo? = null,
        val writechars: Int? = null,
        val hangable: YesNo? = null,
        val holdsliquid: YesNo? = null,
        val mana: Int? = null,
        val damagetype: DamageElement? = null,
        val damage: String? = null,
        val volume: Int? = null,
        val duration: String? = null,
        val destructible: YesNo? = null,
        val droppedby: MutableList<String>? = mutableListOf(),
        val value: String? = null,
        val npcvalue: String? = null,
        val npcprice: String? = null,
        val npcvaluerook: String? = null,
        val npcpricerook: String? = null,
        val buyfrom: String? = null,
        val sellto: String? = null,
        val fansite: String? = null
) : WikiObject() {

    override fun setDefaultValues() {
        if (isEmpty(implemented)) {
            this.copy(implemented = "?")
        }
        if (isEmpty(itemclass)) {
            logger.warn("Creature '{}' has no itemclass set", name)
        }
        if (isEmpty(primarytype)) {
            this.copy(primarytype = "?")
            logger.warn("Creature '{}' has no primarytype set", name)
        }
        if (isEmpty(value)) {
            this.copy(value = "?")
        }
        if (isEmpty(npcvalue)) {
            this.copy(npcvalue = "?")
        }
        if (isEmpty(npcprice)) {
            this.copy(npcprice = "?")
        }
        if (isEmpty(buyfrom)) {
            this.copy(buyfrom = "?")
        }
        if (isEmpty(sellto)) {
            this.copy(sellto = "?")
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(Item::class.java)
    }
}